### PR TITLE
Enable mcp2515 CAN controller for intel aero

### PR DESCRIPTION
**Pull request by sugnanprabhu**
Steps to test:

Connect two aero boards over CAN interface
aero1 CAN_L -> aero2 CAN_L
aero1 CAN_H -> aero2 CAN_H
aero1 CAN_GND -> aero2 CAN_GND

Aero1
```
# ip link set can0 type can bitrate 125000 triple-sampling on
# ifconfig can0 up
# candump can0
```
Aero2
```
# ip link set can0 type can bitrate 125000 triple-sampling on
# ifconfig can0 up
# cansend can0 5A1#11.22.33.44.55.66.77.88 
```

### CODE REVIEW

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+diff --git a/drivers/net/can/spi/mcp251x.c b/drivers/net/can/spi/mcp251x.c
+index 575790e..3f68400 100644
+--- a/drivers/net/can/spi/mcp251x.c
++++ b/drivers/net/can/spi/mcp251x.c
+@@ -75,6 +75,7 @@
+ #include <linux/slab.h>
+ #include <linux/spi/spi.h>
+ #include <linux/uaccess.h>
++#include <linux/acpi.h>
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: Linux kernel mandates includes to be sorted. Also make sure to run `./scripts/checkpatch.pl` in the kernel tree and make sure it passes

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ /* SPI interface instruction set */
+@@ -219,6 +220,12 @@
+ 
+ #define DEVICE_NAME "mcp251x"
+ 
++#define CAN_SPI_BUS	1
++#define CAN_SPI_CS	0
++#define CAN_SPI_MZ	(1000000)
++#define OST_FREQ 25*1000*1000
++#define CPU_IN_MCP2515_IRQ 157
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: These need to be a separate patch for the kernel. In summary you need:
1. One patch to add support for acpi
1. One patch to add the workaround for the wrong configuration passed by the BIOS

Next time we update the BIOS we can request this change and drop this patch.

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+@@ -1033,12 +1040,22 @@ static int mcp251x_can_probe(struct spi_device *spi)
+ 	struct clk *clk;
+ 	int freq, ret;
+ 
++	spi->chip_select = CAN_SPI_CS;
++	spi->max_speed_hz = CAN_SPI_MZ;
++	spi->mode = SPI_MODE_0;
++
++	/* spi->irq = acpi_dev_gpio_irq_get(ACPI_COMPANION(&spi->dev), 0); */
++	spi->irq = CPU_IN_MCP2515_IRQ;
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: What are the values coming from the ACPI table? Just override those that are wrong.
Explain in the commit message for the patch with the override the reason for each  override, including how the magic value of 157 was found.

Comment (lucasdemarchi) [reply recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: Also, surround the override with `if (has_acpi_companion()) { }` like is done everywhere else to maintain compatibility between ACPI and OF.

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ 	clk = devm_clk_get(&spi->dev, NULL);
+ 	if (IS_ERR(clk)) {
+-		if (pdata)
++		if (pdata) {
+ 			freq = pdata->oscillator_frequency;
+-		else
+-			return PTR_ERR(clk);
++		} else {
++			/* pr_err("clock error");
++			return PTR_ERR(clk); */
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: Never add commented out code

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
++			freq = OST_FREQ;
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: This should be an override, together with the others. Ideally on the acpi probe you would set the value in a pdata structure so you don't have to do the if/else here.

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
++		}
+ 	} else {
+ 		freq = clk_get_rate(clk);
+ 	}
+@@ -1067,10 +1084,13 @@ static int mcp251x_can_probe(struct spi_device *spi)
+ 	priv->can.clock.freq = freq / 2;
+ 	priv->can.ctrlmode_supported = CAN_CTRLMODE_3_SAMPLES |
+ 		CAN_CTRLMODE_LOOPBACK | CAN_CTRLMODE_LISTENONLY;
++#if 0
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: Rework the patch so you don't need hacks like this. Don't leave commented out code

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
++	{ },
++};
++MODULE_DEVICE_TABLE(acpi, pxa2xx_spi_acpi_match);
++
+ static struct spi_driver mcp251x_can_driver = {
+ 	.driver = {
+ 		.name = DEVICE_NAME,
+ 		.of_match_table = mcp251x_of_match,
+ 		.pm = &mcp251x_can_pm_ops,
++		.acpi_match_table = ACPI_PTR(pxa2xx_spi_acpi_match),
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: acpi_match_table should be moved up, after of_match_table

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
++	dev_info(&spi->dev, "CAN driver registration successful\n");
++
+ 	return 0;
+ 
+ error_probe:
+@@ -1240,11 +1262,18 @@ static int __maybe_unused mcp251x_can_resume(struct device *dev)
+ static SIMPLE_DEV_PM_OPS(mcp251x_can_pm_ops, mcp251x_can_suspend,
+ 	mcp251x_can_resume);
+ 
++static const struct acpi_device_id pxa2xx_spi_acpi_match[] = {
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: wrong name for the table... this is mcp2515_acpi_match

**Review by lucasdemarchi — CHANGES_REQUESTED** [review]
The summary of the review is that the addition of the ACPI table needs to be done in a clear way that may be sent to upstream.  Hacks to our platform needs to be done in a separate PR.

**Comment by lucasdemarchi** [comment]
Commit messages should always have verbs in imperative, not gerund. So it's "Enable mcp2515..." not "Enabling mcp2515..."

File: recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch (hunk 2 above)
Comment (zehortigoza) [reply recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: Or have a ACPI overlay instead of the workaround patch.

Comment (lucasdemarchi) [reply recipes-kernel/linux/linux-yocto/0045-MCP2515-probe-using-ACPI-table-entry.patch]: even better... I'm not sure you can replace a device with an overlay though. I think the situation (at least on 4.4) is that you can add stuff, not replace. I may be wrong.

**Comment by lucasdemarchi** [comment]
ping

**Comment by sugnanprabhu** [comment]
@lucasdemarchi started working on the changes, exploring if we can override the ACPI values with overlay.
Following is one of the approach found, but it is only for debug.
https://01.org/linux-acpi/documentation/overriding-dsdt

Let me know if we can proceed with having a patch to add the workaround for the wrong configuration passed by the BIOS for now, and replace it with the override if it works later.

**Comment by sugnanprabhu** [comment]
@lucasdemarchi I have separated out the ACPI changes from the hardcoded values required for the CAN to work on aero. I will submit a new PR once this PR is accepted.

File: recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+ 
++#include <linux/acpi.h>
+ #include <linux/can/core.h>
+ #include <linux/can/dev.h>
+ #include <linux/can/led.h>
+@@ -219,6 +220,8 @@
+ 
+ #define DEVICE_NAME "mcp251x"
+ 
++#define DEFAULT_OSC_CLK_FREQ	(25 * 1000 * 1000)
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch]: this should not be here

File: recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
++		if (spi->irq < 0)
++			return -EINVAL;
++	}
+ 
+ 	clk = devm_clk_get(&spi->dev, NULL);
+ 	if (IS_ERR(clk)) {
+ 		if (pdata)
+ 			freq = pdata->oscillator_frequency;
++		else if (has_acpi_companion(&spi->dev))
++			freq = DEFAULT_OSC_CLK_FREQ;
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch]: this should be part of the "hardcoded-values-patch".

File: recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+ 		priv->model = spi_get_device_id(spi)->driver_data;
++
+ 	priv->net = net;
+ 	priv->clk = clk;
+ 
+@@ -1155,6 +1180,8 @@ static int mcp251x_can_probe(struct spi_device *spi)
+ 
+ 	devm_can_led_init(net);
+ 
++	dev_info(&spi->dev, "CAN driver registration successful\n");
```
Comment (lucasdemarchi) [inline recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch]: this is not part of a "add acpi" patch

**Comment by lucasdemarchi** [comment]
> @lucasdemarchi I have separated out the ACPI changes from the hardcoded values required for the CAN to work on aero. I will submit a new PR once this PR is accepted.

can you submit the patches on this PR? You don't need to create a PR for each commit and it makes sense to have them here.

File: recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch (hunk 10 above)
Comment (sugnanprabhu) [reply recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch]: Sure I will move it.

File: recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch (hunk 11 above)
Comment (sugnanprabhu) [reply recipes-kernel/linux/linux-yocto/0051-Add-ACPI-support-for-MCP251x-CAN-controller.patch]: Shall I make this part of the hardcode patch? Sometimes even though the driver is loaded, can network interface will not be registered?

**Comment by sugnanprabhu** [comment]
> Hacks to our platform needs to be done in a separate PR.

Sure, I was confused by the above comment. I will push the hardcoded values as a different commit on the same PR.

**Comment by zehortigoza** [comment]
This broke the SPI interface with FPGA.

dmesg:
[    3.922844] SPI master driver not found...return value 0
[    3.935413] systemd-journald[144]: Received request to flush runtime journal from PID 1
[    3.940423] SPI master driver not found...return value 0

root@intel-aero:~# lsmod | grep spi
spi_imu                 1136  0

**Comment by sugnanprabhu** [comment]
@zehortigoza I tried the latest synced repo and tested. I do not face any such issue.

root@intel-aero:~# lsmod | grep spi
spi_imu                 1136  0
spi_fpga                1201  0

dmesg:
[    3.886150] spidev fpga registered
[    3.888215] spidev-imu: spidev registered

root@intel-aero:~# lsmod | grep mcp
mcp251x                11857  0
can_dev                 9665  1 mcp251x

aero-get-version.py:
BIOS_VERSION = Aero-01.00.13
OS_VERSION = Poky Aero (Intel Aero Linux Distro) 1.6.0-dev (pyro)"
AIRMAP_VERSION = 1.8
FPGA_VERSION = 0xc1

I rebooted the device some 5-10 times and did not face the issues you mentioned and I'm able to flash jam files.

How frequently are you facing this issue?
